### ISSUE TITLE
Change Request: Isolated configs that don't merge with others

### ISSUE THREAD

**Issue by mnaoumov** [eslint/eslint#20354]
### ESLint version

v9.39.1

### What problem do you want to solve?

Originally reported as bug #20352 

There are situations when you need to define a config, but you don't want this config to be merged with other configs. Especially if you want to extract this problematic config into its own plugin, and consumers of this plugin should not need to adjust their configs in order to enable the plugin.

The use case is in the bug mentioned above.

### What do you think is the correct solution?

Additional config flag such as `isolated: true` to prevent the config to be merged with other configs

### Participation

- [ ] I am willing to submit a pull request for this change.

**Comment by nzakas** [comment]
Sorry, I'm having a hard time imagining what this use case could be. Can you explain with something concrete?

**Comment by mnaoumov** [comment]
@nzakas 

Consider the following [ESLint plugin](https://github.com/obsidianmd/eslint-plugin/blob/903401e/lib/index.ts)

Here we have mostly `TypeScript` rules but [also](https://github.com/obsidianmd/eslint-plugin/blob/903401e/lib/index.ts#L244-L260) we have some rules that apply exclusively to `package.json`

As you see in [L245](https://github.com/obsidianmd/eslint-plugin/blob/903401e/lib/index.ts#L245) we have to add `files: ['package.json']`

But then this `files` configuration is getting merged into other configurations, which potentially breaks them

We would like our plugin to be used by other projects in the way

```js
export default defineConfig([
  userConfig1,
  userConfig2,
  ...obsidianmd.configs.recommended
]);
```

but as `files` configuration gets merged into all configs, especially the users one that is beyond our control, it can introduce bugs, if `userConfig` doesn't handle `package.json` well, e.g.

```
Error: Error while loading rule '@typescript-eslint/await-thenable': You have used a rule which requires type information, but don't have parserOptions set to generate type information for this file. See https://typescript-eslint.io/getting-started/typed-linting for enabling linting with type information.
Parser: typescript-eslint/parser
Occurred while linting F:\foo\package.json
```

This could be worked around by adding `ignores: ['package.json']` into every single config where such merge was undesired.

However this is clearly a hack that requires our plugin users to modify configs on their end.

Clearly it breaks the idea of encapsulation.

If we add `ignores: ['package.json']` to our plugin, the error goes away, but the [json rules](https://github.com/obsidianmd/eslint-plugin/blob/903401e/lib/index.ts#L244-L260) no longer applied.

I could not find a way to provide working configuration that our plugin's users can just plug-in without fixing their own configs.

That's why I originally raised this as an ESLint bug #20352

**Comment by JoshuaKGoldberg** [comment]
>  if `userConfig` doesn't handle `package.json` well

I think that's the root of your issue. If `userConfig` can't handle a certain type of files, it shouldn't allow itself to be run on those files. See #20075 -> #20158: ESLint's docs now recommend explicitly stating the `files` for each area of rules. ACK that you do have that set up already for `.json`, etc. (great!) - but it looks like `js.configs.recommended` is configured for all rules.

Unless, am I misreading? Why would e.g. `await-thenable` be applying to `.json` files if not because of a missing `files` config value?

**Comment by mnaoumov** [comment]
@JoshuaKGoldberg the problem is that it's currently impossible to express the fact that `json` rules should be applied to `package.json` file only without leaking into other rules that were not designed to deal with it

**Comment by JoshuaKGoldberg** [comment]
Is it? I use a pattern like this in a lot of my configs:

```js
export default defineConfig(
  {
    files: ["**/package.json"],
    extends: [packageJson.configs.recommended],
  },
  {
    files: ["**/*.json"],
    extends: [jsonc.configs.recommended],
  }
);
```

If I'm misinterpreting what you're saying, it might be because there's a lot of code to read through in the provided configs. Would you be able to provide a smaller reproduction that'd be easier to read through?

**Comment by mnaoumov** [comment]
Simplified example

`eslint.config.mts`

```ts
import { myPluginConfig } from './myPlugin.mts';

export default defineConfig([
  {
    files: ['src/**/*.ts'],
  },
  {
    rules: [someRuleWorksOnlyForTs] // fails as it applies to package.json
  },
  myPluginConfig
]);
```

`myPlugin.mts`

```ts
export const myPluginConfig = defineConfig([
  {
    files: ['package.json'];
    rules: [someRuleThatWorksOnlyForPackageJson] // works fine
  },
  {
    files: ['**/*.ts'];
    rules: [someOtherRuleWorksOnlyForTs] // works fine
  },
]);
```

**Comment by JoshuaKGoldberg** [comment]
That first code block isn't how flat config works. Putting `files` alone in a config object ~doesn't do anything~ _edit:_ adds them for all lint rules later on. What you actually want is:

```js
import { myPluginConfig } from './myPlugin.mts';

export default defineConfig([
  {
    files: ['src/**/*.ts'],
    rules: [someRuleWorksOnlyForTs],
  },
  myPluginConfig
]);
```

~💡 I wonder if `defineConfig` should emit some kind of warning for config objects that do nothing...?~

**Comment by mnaoumov** [comment]
Updated example (previous one was incorrect)

`eslint.config.mts`

```ts
import { myPluginConfig } from './myPlugin.mts';

export default defineConfig([
  {
    files: ['src/**/*.ts'],
    rules: [someRuleWorksOnlyForTs] // fails as it applies to package.json, adding workaround as below would help, but it breaks configs encapsulation, so that's the reason for the FR
  },
  myPluginConfig
]);
```

`myPlugin.mts`

```ts
export const myPluginConfig = defineConfig([
  {
    files: ['package.json'];
    rules: [someRuleThatWorksOnlyForPackageJson] // works fine
  },
  {
    files: ['**/*.ts'],
    ignores: ['package.json'], // hacky workaround
    rules: [someOtherRuleWorksOnlyForTs] // works fine
  },
]);
```

**Comment by JoshuaKGoldberg** [comment]
> ```ts
> files: ['src/**/*.ts'],
> rules: [someRuleWorksOnlyForTs] // fails as it applies to package.json
> ```

I don't see how `someRuleWorksOnlyForTs` would still apply to `package.json` in that case. The config object is explicitly limiting it to only apply to `src/**/*.ts` files.

Could you please post a [minimum reproduction project](https://antfu.me/posts/why-reproductions-are-required#reproducible-projects-or-playgrounds) that we can quickly get up and running to see what you mean? My hunch is by making that you'll see some other issue that's not this one - but hey, maybe I'm wrong?

**Comment by mnaoumov** [comment]
Created fully working example

https://github.com/mnaoumov/eslint-isolated-config-example

**Comment by mnaoumov** [comment]
@JoshuaKGoldberg 
One of the examples I prepared on the link above (`v3`) works as expected but it indeed requires adding `files` to each config as per links you provided above.

**Comment by JoshuaKGoldberg** [comment]
Yup, same as before, you are defining `files` on its own. That ~does nothing~ adds more files to be linted generally. https://github.com/mnaoumov/eslint-isolated-config-example/blob/6efb88a118c25d31c60c30c707fa7f248f8eed69/eslint.config.mts#L6-L11:

```ts
export default defineConfig([
  {
    files: ['./src/*.ts']
  },
  {
    plugins: { otherPlugin },
```

Removing the `},\n  {` stops `someOtherRuleThatWorksOnlyForTs` from operating on non-`./src/*.ts` files.

At this point, I'm pretty convinced this is not an issue with ESLint. Isolating some rules to only work on some extensions is a primary use case of modern ESLint configs. It'd be very surprising if the use cases mentioned in this issue didn't work!

Closing out this issue as more of a support ticket, which is typically handled in the [ESLint Discord](https://eslint.org/chat). But if I'm wrong, we can always re-open.

**Comment by kirkwaiblinger** [comment]
`files` (whether the rest of the config is object empty or not) will also add files to the lint run though if they're not already part of it... https://eslint.org/docs/latest/use/configure/configuration-files#specifying-files-and-ignores

**Comment by kirkwaiblinger** [comment]
Repro to play with: https://github.com/kirkwaiblinger/repro-repo/tree/eslint-eslint-20354

Related issues: 
- https://github.com/typescript-eslint/typescript-eslint/issues/11300 (adding linting for other languages breaks users' working JS/TS configs)
- https://github.com/typescript-eslint/typescript-eslint/issues/11291 (typescript-eslint configs are mostly unscoped)

**Comment by JoshuaKGoldberg** [comment]
Oh, wow, I don't remember knowing that. Scary. Thanks Kirk - edited my posts.
